File: swift/Sources/Wallet/FollowUpContext.swift
import Foundation
import OpenID4VCI
import SdJwt
import WalletAPI

/// Everything needed to resume a credential after issuance — deferred poll, reissue, notification —
/// across app restarts (assembly gap #3). Key material stays in the SecureArea; only handles serialize.
struct FollowUpContext {
    let credentialIssuer: String
    let configurationId: String
    let requestedFormat: String
    let accessToken: String?
    let refreshToken: String?
    let transactionId: String?
    let notificationId: String?
    let proofKeys: [KeyHandle]
    let dpopKey: KeyHandle
    let policy: CredentialPolicy

    /// A copy with a fresh `transaction_id` — the issuer rotates it on each still-deferred (202) response.
    func withTransactionId(_ transactionId: String?) -> FollowUpContext {
        FollowUpContext(credentialIssuer: credentialIssuer, configurationId: configurationId, requestedFormat: requestedFormat,
                        accessToken: accessToken, refreshToken: refreshToken, transactionId: transactionId,
                        notificationId: notificationId, proofKeys: proofKeys, dpopKey: dpopKey, policy: policy)
    }

    /// Reconstructs the vci `CredentialResponse` carrying the follow-up context (public init).
    func toCredentialResponse() -> CredentialResponse {
        CredentialResponse(credentials: [], transactionId: transactionId, notificationId: notificationId,
                           accessToken: accessToken, credentialIssuer: credentialIssuer, requestedFormat: requestedFormat,
                           refreshToken: refreshToken, configurationId: configurationId)
    }

    func encode() -> [UInt8] {
        let json = JsonValue.obj([
            ("credentialIssuer", .str(credentialIssuer)),
            ("configurationId", .str(configurationId)),
            ("requestedFormat", .str(requestedFormat)),
            ("accessToken", accessToken.map(JsonValue.str) ?? .null),
            ("refreshToken", refreshToken.map(JsonValue.str) ?? .null),
            ("transactionId", transactionId.map(JsonValue.str) ?? .null),
            ("notificationId", notificationId.map(JsonValue.str) ?? .null),
            ("proofKeys", .arr(proofKeys.map(handleJson))),
            ("dpopKey", handleJson(dpopKey)),
            ("batchSize", .numInt(Int64(policy.batchSize))),
            ("use", .str(policy.use == .rotate ? "rotate" : "oneTime")),
        ])
        return Array(json.serialize().utf8)
    }

    static func decode(_ bytes: [UInt8]) throws -> FollowUpContext {
        let o = try JsonValue.parse(String(decoding: bytes, as: UTF8.self))
        func str(_ k: String) -> String { if case let .str(s)? = o[k] { return s }; return "" }
        func strOrNil(_ k: String) -> String? { if case let .str(s)? = o[k] { return s }; return nil }
        var proofs: [KeyHandle] = []
        if case let .arr(items)? = o["proofKeys"] { proofs = items.map(keyHandle) }
        var batch: Int64 = 1
        if case let .numInt(n)? = o["batchSize"] { batch = n }
        return FollowUpContext(
            credentialIssuer: str("credentialIssuer"), configurationId: str("configurationId"),
            requestedFormat: str("requestedFormat"), accessToken: strOrNil("accessToken"),
            refreshToken: strOrNil("refreshToken"), transactionId: strOrNil("transactionId"),
            notificationId: strOrNil("notificationId"), proofKeys: proofs, dpopKey: keyHandle(o["dpopKey"] ?? .null),
            policy: CredentialPolicy(batchSize: Int(batch), use: str("use") == "oneTime" ? .oneTime : .rotate))
    }
}

private func handleJson(_ h: KeyHandle) -> JsonValue {
    .obj([("secureArea", .str(h.secureArea.value)), ("alias", .str(h.alias))])
}

private func keyHandle(_ v: JsonValue) -> KeyHandle {
    func s(_ k: String) -> String { if case let .str(x)? = v[k] { return x }; return "" }
    return KeyHandle(secureArea: SecureAreaId(s("secureArea")), alias: s("alias"))
}

File: kotlin/wallet/src/main/kotlin/com/hopae/eudi/wallet/FollowUpContext.kt
package com.hopae.eudi.wallet

import com.hopae.eudi.wallet.sdjwt.JsonValue
import com.hopae.eudi.wallet.spi.CredentialPolicy
import com.hopae.eudi.wallet.spi.KeyHandle
import com.hopae.eudi.wallet.spi.KeyUse
import com.hopae.eudi.wallet.spi.SecureAreaId
import com.hopae.eudi.wallet.vci.CredentialResponse

/**
 * Everything needed to resume a credential after issuance — deferred poll, reissue, notification —
 * across app restarts (assembly gap #3). Persisted as JSON in the envelope (deferred) or a storage
 * collection (reissue). Key material stays in the SecureArea; only handles are serialized.
 */
internal class FollowUpContext(
    val credentialIssuer: String,
    val configurationId: String,
    val requestedFormat: String,
    val accessToken: String?,
    val refreshToken: String?,
    val transactionId: String?,
    val notificationId: String?,
    val proofKeys: List<KeyHandle>,
    val dpopKey: KeyHandle,
    val policy: CredentialPolicy,
) {
    /** A copy with a fresh `transaction_id` — the issuer rotates it on each still-deferred (202) response. */
    fun withTransactionId(transactionId: String?): FollowUpContext = FollowUpContext(
        credentialIssuer, configurationId, requestedFormat, accessToken, refreshToken,
        transactionId, notificationId, proofKeys, dpopKey, policy,
    )

    /** Reconstructs the vci [CredentialResponse] carrying the follow-up context (public ctor, no `withContext`). */
    fun toCredentialResponse(): CredentialResponse = CredentialResponse(
        credentials = emptyList(),
        transactionId = transactionId,
        notificationId = notificationId,
        accessToken = accessToken,
        credentialIssuer = credentialIssuer,
        requestedFormat = requestedFormat,
        refreshToken = refreshToken,
        configurationId = configurationId,
    )

    fun encode(): ByteArray = JsonValue.Obj(
        listOf(
            "credentialIssuer" to JsonValue.Str(credentialIssuer),
            "configurationId" to JsonValue.Str(configurationId),
            "requestedFormat" to JsonValue.Str(requestedFormat),
            "accessToken" to nullable(accessToken),
            "refreshToken" to nullable(refreshToken),
            "transactionId" to nullable(transactionId),
            "notificationId" to nullable(notificationId),
            "proofKeys" to JsonValue.Arr(proofKeys.map { it.toJson() }),
            "dpopKey" to dpopKey.toJson(),
            "batchSize" to JsonValue.NumInt(policy.batchSize.toLong()),
            "use" to JsonValue.Str(policy.use.name),
        ),
    ).serialize().encodeToByteArray()

    companion object {
        fun decode(bytes: ByteArray): FollowUpContext {
            val o = JsonValue.parse(bytes.decodeToString()) as JsonValue.Obj
            return FollowUpContext(
                credentialIssuer = o.str("credentialIssuer"),
                configurationId = o.str("configurationId"),
                requestedFormat = o.str("requestedFormat"),
                accessToken = o.strOrNull("accessToken"),
                refreshToken = o.strOrNull("refreshToken"),
                transactionId = o.strOrNull("transactionId"),
                notificationId = o.strOrNull("notificationId"),
                proofKeys = (o["proofKeys"] as JsonValue.Arr).items.map { keyHandle(it as JsonValue.Obj) },
                dpopKey = keyHandle(o["dpopKey"] as JsonValue.Obj),
                policy = CredentialPolicy((o["batchSize"] as JsonValue.NumInt).value.toInt(), KeyUse.valueOf(o.str("use"))),
            )
        }
    }
}

private fun nullable(value: String?): JsonValue = value?.let { JsonValue.Str(it) } ?: JsonValue.Null
private fun JsonValue.Obj.str(key: String): String = (this[key] as JsonValue.Str).value
private fun JsonValue.Obj.strOrNull(key: String): String? = (this[key] as? JsonValue.Str)?.value
private fun KeyHandle.toJson(): JsonValue = JsonValue.Obj(listOf("secureArea" to JsonValue.Str(secureArea.value), "alias" to JsonValue.Str(alias)))
private fun keyHandle(o: JsonValue.Obj): KeyHandle = KeyHandle(SecureAreaId(o.str("secureArea")), o.str("alias"))
